tests/python/relay/test_pass_plan_devices.py

```diff
@@ -0,0 +1,1405 @@
+        mod = tvm.IRModule()
+        mod[f] = relay.Function(
+            [a, b], relay.add(a, on_cpu(b)), relay.ty.TensorType(shape, "float32")
+        )
+        mod[main] = relay.Function(
+            [x, y], relay.Call(f, [y, x]), relay.ty.TensorType(shape, "float32")
+        )
+        return mod
+
+    # def @f(a, b, on_device={param_device_types=[2,1], result_device_type=2}) { add(a, on_cpu(b)) }
+    # def @main(x, y, on_device={param_device_types=[1,2], result_device_type=2}) { @f(y, x) }
+    def expected():
+        mod = tvm.IRModule()
+        mod[f] = relay.annotation.function_on_device(
+            relay.Function(
+                [a, b], relay.add(a, cpu_to_gpu(b)), relay.ty.TensorType(shape, "float32")
+            ),
+            [GPU, CPU],
+            GPU,
+        )
+        mod[main] = relay.annotation.function_on_device(
+            relay.Function([x, y], relay.Call(f, [y, x]), relay.ty.TensorType(shape, "float32")),
+            [CPU, GPU],
+            GPU,
+        )
+        return mod
+
+    def ref(x, y):
+        def f(a, b):
+            return np.add(a, b)
+
+        return f(x, y)
+
+    exercise(input(), expected(), ref, rands(shape, 2))
+
+
+# Note that match and ADTs don't appear to be supported for direct AST
+# construction.
+
+
+def test_ref():
+    shape = (N, M)
+    x = relay.var("x", shape=shape)
+    y = relay.var("y", shape=shape)
+    r = relay.var("r")
+    dummy = relay.var("dummy")
+
+    # def @main(x, y) {
+    #   r = ref(x)
+    #   ref_write(r, on_cpu(y))
+    #   add(x, ref_read(r))
+    # }
+    def input():
+        return tvm.IRModule.from_expr(
+            relay.Function(
+                [x, y],
+                relay.Let(
+                    r,
+                    relay.RefCreate(x),
+                    relay.Let(dummy, relay.RefWrite(r, on_cpu(y)), relay.add(x, relay.RefRead(r))),
+                ),
+            )
+        )
+
+    # def @main(x, y, on_device={param_device_types=[GPU, CPU], result_device_type=GPU}) {
+    #   r = ref(x)
+    #   ref_write(r, cpu_to_gpu(y))
+    #   add(x, ref_read(r))
+    # }
+    def expected():
+        return tvm.IRModule.from_expr(
+            relay.annotation.function_on_device(
+                relay.Function(
+                    [x, y],
+                    relay.Let(
+                        r,
+                        relay.RefCreate(x),
+                        relay.Let(
+                            dummy, relay.RefWrite(r, cpu_to_gpu(y)), relay.add(x, relay.RefRead(r))
+                        ),
+                    ),
+                ),
+                [GPU, CPU],
+                GPU,
+            )
+        )
+
+    def ref(x, y):
+        r = {"value": x}
+        r["value"] = y
+        return np.add(x, r["value"])
+
+    # Don't try to execute, no backend currently supports both cross-devices and references.
+    exercise(input(), expected(), None, None)
+
+
+if __name__ == "__main__":
+    test_plain()
```

Comment:
`sys.exit(pytest.main([__file__] + sys.argv[1:]))`

one day we will write a helper :)

Reply:
ping pong on this one

Reply:
done